nano/node/active_transactions.cpp

```diff
@@ -717,85 +749,88 @@ void nano::active_transactions::update_adjusted_difficulty ()
+					roots.get<tag_root> ().modify (existing_root, [multiplier_a](nano::conflict_info & info_a) {
+						info_a.adjusted_multiplier = multiplier_a;
 					});
 				}
 			}
 		}
 	}
 }
 
-void nano::active_transactions::update_active_difficulty (nano::unique_lock<std::mutex> & lock_a)
+void nano::active_transactions::update_active_multiplier (nano::unique_lock<std::mutex> & lock_a)
 {
 	debug_assert (!mutex.try_lock ());
-	last_prioritized_difficulty.reset ();
+	last_prioritized_multiplier.reset ();
 	double multiplier (1.);
 	// Heurestic to filter out non-saturated network and frontier confirmation
 	if (roots.size () > prioritized_cutoff / 2 || (node.network_params.network.is_test_network () && !roots.empty ()))
 	{
 		auto & sorted_roots = roots.get<tag_difficulty> ();
-		std::vector<uint64_t> prioritized;
+		std::vector<double> prioritized;
 		prioritized.reserve (std::min (sorted_roots.size (), prioritized_cutoff));
 		for (auto it (sorted_roots.begin ()), end (sorted_roots.end ()); it != end && prioritized.size () < prioritized_cutoff; ++it)
 		{
 			if (!it->election->confirmed ())
 			{
-				prioritized.push_back (it->adjusted_difficulty);
+				prioritized.push_back (it->adjusted_multiplier);
 			}
 		}
 		if (prioritized.size () > 10 || (node.network_params.network.is_test_network () && !prioritized.empty ()))
 		{
-			multiplier = nano::difficulty::to_multiplier (prioritized[prioritized.size () / 2], node.network_params.network.publish_thresholds.epoch_1);
+			multiplier = prioritized[prioritized.size () / 2];
 		}
 		if (!prioritized.empty ())
 		{
-			last_prioritized_difficulty = prioritized.back ();
+			last_prioritized_multiplier = prioritized.back ();
 		}
 	}
 	debug_assert (multiplier >= nano::difficulty::to_multiplier (node.network_params.network.publish_thresholds.entry, node.network_params.network.publish_thresholds.epoch_1));
 	multipliers_cb.push_front (multiplier);
 	auto sum (std::accumulate (multipliers_cb.begin (), multipliers_cb.end (), double(0)));
-	auto difficulty = nano::difficulty::from_multiplier (sum / multipliers_cb.size (), node.network_params.network.publish_thresholds.epoch_1);
+	double avg_multiplier (sum / multipliers_cb.size ());
+	auto difficulty = nano::difficulty::from_multiplier (avg_multiplier, node.default_difficulty (nano::work_version::work_1));
 	debug_assert (difficulty >= node.network_params.network.publish_thresholds.entry);
 
-	trended_active_difficulty = difficulty;
-	node.observers.difficulty.notify (trended_active_difficulty);
+	trended_active_multiplier = avg_multiplier;
+	node.observers.difficulty.notify (difficulty);
 }
 
 uint64_t nano::active_transactions::active_difficulty ()
 {
-	nano::lock_guard<std::mutex> lock (mutex);
-	return trended_active_difficulty;
+	return nano::difficulty::from_multiplier (active_multiplier (), node.default_difficulty (nano::work_version::work_1));
+}
+
+uint64_t nano::active_transactions::limited_active_difficulty (std::shared_ptr<nano::block> block_a)
```

Comment:
This could be a `const & nano::block` instead. Similar for `normalized_multiplier`.

Reply:
Updated